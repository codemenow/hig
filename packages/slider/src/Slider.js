import React, { Component } from "react";
import PropTypes from "prop-types";
import { css } from "emotion";

import Input from "./presenters/Input";
import stylesheet from "./Slider.stylesheet";

/**
 * @typedef {string|number} Value
 */

/**
 * @typedef {Object} State
 * @property {Value} value
 */

export default class Slider extends Component {
  static propTypes = {
    /**
     * Initial value of the field. User action will override
     */
    defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /**
     * Prevents user actions on the field
     */
    disabled: PropTypes.bool,
    /**
     * HTML ID attribute
     */
    id: PropTypes.string,
    /**
     * Minimum value of the slider
     */
    min: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /**
     * Maximum value of the slider
     */
    max: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /**
     * Name of the field when submitted with a form
     */
    name: PropTypes.string,
    /**
     * Called when user moves focus from the field
     */
    onBlur: PropTypes.func,
    /**
     * Called after user changes the value of the field
     */
    onChange: PropTypes.func,
    /**
     * Called when user puts focus onto the field
     */
    onFocus: PropTypes.func,
    /**
     * Called as user changes the value of the field
     */
    onInput: PropTypes.func,
    /**
     * The granularity of each step on the slider
     */
    step: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /**
     * Value of the field
     */
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  };

  static defaultProps = {
    defaultValue: "",
    min: "0",
    max: "100",
    step: "1"
  };

  /**
   * @type {State}
   */
  state = {
    value: this.props.defaultValue
  };

  /**
   * @returns {Value}
   */
  getValue() {
    if (this.isControlled()) {
      return this.props.value;
    }

    return this.state.value;
  }

  /**
   * @param {Value} value
   */
  setValue(value) {
    const { onChange } = this.props;

    if (onChange) onChange(Number(value));

    if (!this.isControlled()) {
      this.setState({ value });
    }
  }

  isControlled() {
    return this.props.value !== undefined;
  }

  /**
   * @param {event} Event
   */
  handleChange = event => {
    this.setValue(event.target.value);
  };

  render() {
    const {
      defaultValue, // exclude from otherProps
      onChange, // exclude from otherProps
      ...otherProps
    } = this.props;

    const value = this.getValue();
    const styles = stylesheet();

    return (
      <div className={css(styles.slider)}>
        <Input onChange={this.handleChange} value={value} {...otherProps} />
      </div>
    );
  }
}
