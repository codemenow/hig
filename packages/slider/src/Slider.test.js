import React from "react";
import { takeSnapshotsOf } from "@hig/jest-preset/helpers";
import { mount } from "enzyme";
import Slider from "./index";
import Input from "./presenters/Input";

describe("slider/Slider", () => {
  takeSnapshotsOf(Slider, [{ desc: "renders", props: {} }]);
  takeSnapshotsOf(Slider, [
    { desc: "renders disabled", props: { disabled: true } }
  ]);

  describe("when uncontrolled", () => {
    it("calls the `onChange`the changed value", () => {
      const eventHandler = jest.fn();
      const wrapper = mount(<Slider onChange={eventHandler} />);
      wrapper.find("input").simulate("change", { target: { value: "10" } });

      expect(eventHandler).toHaveBeenCalledWith(10);
    });

    it("allows value to be changed by change events", () => {
      const wrapper = mount(<Slider defaultValue="10" />);
      const input = wrapper.find("input");

      expect(input.prop("value")).toEqual("10");
      input.simulate("change", { target: { value: "20" } });
      expect(input.prop("value")).toEqual("20");
    });
  });

  describe("when controlled", () => {
    it("calls the `onChange` handler", () => {
      const eventHandler = jest.fn();
      const wrapper = mount(<Slider value="8" onChange={eventHandler} />);
      wrapper.find("input").simulate("change", { target: { value: "10" } });

      expect(eventHandler).toHaveBeenCalledWith(10);
    });

    it("doesn't allow the value to be changed by change events", () => {
      const wrapper = mount(<Slider value="10" />);
      const input = wrapper.find("input");

      expect(input.prop("value")).toEqual("10");
      input.simulate("change", { target: { value: "20" } });
      expect(input.prop("value")).toEqual("10");
    });

    it("recognizes changes to the value prop", () => {
      const wrapper = mount(<Slider value="10" />);
      const input = wrapper.find("input");

      expect(input.prop("value")).toEqual("10");
      wrapper.setProps({ value: "20" });
      expect(input.prop("value")).toEqual("20");
    });
  });
});
