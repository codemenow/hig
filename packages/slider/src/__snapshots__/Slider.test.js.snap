// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slider/Slider renders 1`] = `
.emotion-1 {
  display: block;
}

.emotion-0 {
  -webkit-appearance: none;
  width: 100%;
  background-color: transparent;
  box-sizing: content-box;
  height: 20px;
  margin: 0;
  position: relative;
  outline: none;
}

.emotion-0::-ms-tooltip {
  display: none;
}

.emotion-0::-ms-tooltip {
  display: none;
}

.emotion-0::-moz-focus-outer {
  border: 0;
}

.emotion-0::-ms-thumb {
  box-sizing: content-box;
  height: 16px;
  width: 16px;
  background-color: #D4DBE1;
  border: 0;
  border-radius: 50%;
  box-shadow: 0 0 0 2px #ffffff;
  position: relative;
  top: 50%;
  -webkit-appearance: none;
}

.emotion-0::-moz-range-thumb {
  box-sizing: content-box;
  height: 16px;
  width: 16px;
  background-color: #D4DBE1;
  border: 0;
  border-radius: 50%;
  box-shadow: 0 0 0 2px #ffffff;
  position: relative;
  top: 50%;
  -webkit-appearance: none;
}

.emotion-0::-webkit-slider-thumb {
  box-sizing: content-box;
  height: 16px;
  width: 16px;
  background-color: #D4DBE1;
  border: 0;
  border-radius: 50%;
  box-shadow: 0 0 0 2px #ffffff;
  position: relative;
  top: 50%;
  -webkit-appearance: none;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.emotion-0::-ms-track {
  width: 100%;
  height: 2px;
  border: none;
  background-color: rgba(212,219,225,0.5);
  color: transparent;
  outline: none;
}

.emotion-0::-moz-range-track {
  width: 100%;
  height: 2px;
  border: none;
  background-color: rgba(212,219,225,0.5);
  color: transparent;
  outline: none;
}

.emotion-0::-webkit-slider-runnable-track {
  width: 100%;
  height: 2px;
  border: none;
  background-color: rgba(212,219,225,0.5);
  color: transparent;
  outline: none;
  background-image: linear-gradient(#D4DBE1,#D4DBE1);
  background-position: 0;
  background-size: calc((0.5 * 16px)  + (0.5 * (100% - 16px))) 100%;
  background-repeat: no-repeat;
}

.emotion-0::-ms-fill-lower {
  border: none;
  background-color: #D4DBE1;
}

.emotion-0::-moz-range-progress {
  border: none;
  background-color: #D4DBE1;
}

<div
  className="emotion-1"
>
  <input
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow={50}
    className="emotion-0"
    disabled={false}
    max="100"
    min="0"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    step="1"
    type="range"
    value={50}
  />
</div>
`;

exports[`slider/Slider renders disabled 1`] = `
.emotion-1 {
  display: block;
}

.emotion-0 {
  -webkit-appearance: none;
  width: 100%;
  background-color: transparent;
  box-sizing: content-box;
  height: 20px;
  margin: 0;
  position: relative;
  outline: none;
}

.emotion-0::-ms-tooltip {
  display: none;
}

.emotion-0::-ms-tooltip {
  display: none;
}

.emotion-0::-moz-focus-outer {
  border: 0;
}

.emotion-0::-ms-thumb {
  box-sizing: content-box;
  height: 16px;
  width: 16px;
  background-color: #D4DBE1;
  border: 0;
  border-radius: 50%;
  box-shadow: 0 0 0 2px #ffffff;
  position: relative;
  top: 50%;
  -webkit-appearance: none;
}

.emotion-0::-moz-range-thumb {
  box-sizing: content-box;
  height: 16px;
  width: 16px;
  background-color: #D4DBE1;
  border: 0;
  border-radius: 50%;
  box-shadow: 0 0 0 2px #ffffff;
  position: relative;
  top: 50%;
  -webkit-appearance: none;
  opacity: 0.5;
}

.emotion-0::-webkit-slider-thumb {
  box-sizing: content-box;
  height: 16px;
  width: 16px;
  background-color: #D4DBE1;
  border: 0;
  border-radius: 50%;
  box-shadow: 0 0 0 2px #ffffff;
  position: relative;
  top: 50%;
  -webkit-appearance: none;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.emotion-0::-ms-track {
  width: 100%;
  height: 2px;
  border: none;
  background-color: rgba(212,219,225,0.5);
  color: transparent;
  outline: none;
  opacity: 0.5;
}

.emotion-0::-moz-range-track {
  width: 100%;
  height: 2px;
  border: none;
  background-color: rgba(212,219,225,0.5);
  color: transparent;
  outline: none;
  opacity: 0.5;
}

.emotion-0::-webkit-slider-runnable-track {
  width: 100%;
  height: 2px;
  border: none;
  background-color: rgba(212,219,225,0.5);
  color: transparent;
  outline: none;
  opacity: 0.5;
  background-image: linear-gradient(#D4DBE1,#D4DBE1);
  background-position: 0;
  background-size: calc((0.5 * 16px)  + (0.5 * (100% - 16px))) 100%;
  background-repeat: no-repeat;
}

.emotion-0::-ms-fill-lower {
  border: none;
  background-color: #D4DBE1;
}

.emotion-0::-moz-range-progress {
  border: none;
  background-color: #D4DBE1;
  opacity: 0.5;
}

<div
  className="emotion-1"
>
  <input
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow={50}
    className="emotion-0"
    disabled={true}
    max="100"
    min="0"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    step="1"
    type="range"
    value={50}
  />
</div>
`;
